### ISSUE TITLE
Excessive memory allocation with fuzzed file (id3v2)

### ISSUE THREAD

**Issue by tholin** [taglib/taglib#486]
Parsing the following file with the tagreader example will allocate several GB of memory.
https://www.dropbox.com/s/tgjalqpi0gdedr2/taglib_aiff_alloc.aif

This might not technically be a bug because it doesn't hang or crash. But it depletes my 16G of ram.

```
$ ulimit -v 16000000; gdb --args ~/repository/taglib/build/examples/tagreader taglib_aiff_alloc.aif 
GNU gdb (Gentoo 7.7.1 p1) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/cocobo/repository/taglib/build/examples/tagreader...done.
(gdb) r
Starting program: /home/cocobo/repository/taglib/build/examples/tagreader taglib_aiff_alloc.aif
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
******************** "taglib_aiff_alloc.aif" ********************
TagLib: RIFF::File::read() -- Chunk '����' has invalid ID
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc

Program received signal SIGABRT, Aborted.
0x00007ffff6c8a597 in raise () from /lib64/libc.so.6
(gdb) bt full
#0  0x00007ffff6c8a597 in raise () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007ffff6c8b8d8 in abort () from /lib64/libc.so.6
No symbol table info available.
#2  0x00007ffff756e845 in __gnu_cxx::__verbose_terminate_handler() ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#3  0x00007ffff756c976 in __cxxabiv1::__terminate(void (*)()) ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#4  0x00007ffff756c9a3 in std::terminate() ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#5  0x00007ffff756cbc2 in __cxa_throw () from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#6  0x00007ffff756d0ad in operator new(unsigned long) ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#7  0x00007ffff75cba5f in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_Rep::_S_create(unsigned long, unsigned long, std::allocator<wchar_t> const&) ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#8  0x00007ffff75cc67d in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_Rep::_M_clone(std::allocator<wchar_t> const&, unsigned long) ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#9  0x00007ffff75cc724 in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::reserve(unsigned long) () from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#10 0x00007ffff75ccad6 in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::append(unsigned long, wchar_t) () from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.3/libstdc++.so.6
No symbol table info available.
#11 0x00007ffff7a42cde in TagLib::String::copyFromLatin1 (this=0x7fffffffd300, s=0x7fff2bd8f015 "", 
    length=3402287813) at /home/cocobo/repository/taglib/taglib/toolkit/tstring.cpp:743
No locals.
#12 0x00007ffff7a3b3bf in TagLib::String::String (this=0x7fffffffd300, v=..., 
    t=TagLib::String::Latin1) at /home/cocobo/repository/taglib/taglib/toolkit/tstring.cpp:252
No locals.
#13 0x00007ffff791f55c in TagLib::ID3v2::Latin1StringHandler::parse (
    this=0x7ffff7ddaa28 <_ZL20defaultStringHandler>, data=...)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/id3v2tag.cpp:103
No locals.
#14 0x00007ffff797afe6 in TagLib::ID3v2::CommentsFrame::parseFields (this=0x6199e0, data=...)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/frames/commentsframe.cpp:163
        byteAlign = 1
        l = {<TagLib::List<TagLib::ByteVector>> = {
            _vptr.List = 0x7ffff7dce6b0 <vtable for TagLib::ByteVectorList+16>, d = 0x61a040}, 
          d = 0x6199e0}
#15 0x00007ffff797bdea in TagLib::ID3v2::CommentsFrame::CommentsFrame (this=0x6199e0, data=..., 
    h=0x618dd0) at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/frames/commentsframe.cpp:196
No locals.
#16 0x00007ffff7911506 in TagLib::ID3v2::FrameFactory::createFrame (
    this=0x7ffff7dda9c0 <TagLib::ID3v2::FrameFactory::factory>, origData=..., tagHeader=0x618808)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/id3v2framefactory.cpp:187
        f = 0x6184e0
        data = {_vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, static null = {
            _vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, 
            static null = <same as static member of an already seen type>, d = 0x616b40}, 
          d = 0x619b30}
        version = 67
        header = 0x618dd0
        frameID = {_vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, 
          static null = {_vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, 
            static null = <same as static member of an already seen type>, d = 0x616b40}, 
          d = 0x619be0}
#17 0x00007ffff792bd92 in TagLib::ID3v2::Tag::parse (this=0x618680, origData=...)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/id3v2tag.cpp:711
        frame = 0x618c00
---Type <return> to continue, or q <return> to quit---
        data = {_vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, static null = {
            _vptr.ByteVector = 0x7ffff7dce5f0 <vtable for TagLib::ByteVector+16>, 
            static null = <same as static member of an already seen type>, d = 0x616b40}, 
          d = 0x618550}
        frameDataPosition = 88
        frameDataLength = 2114
#18 0x00007ffff792b3e3 in TagLib::ID3v2::Tag::read (this=0x618680)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/id3v2tag.cpp:658
No locals.
#19 0x00007ffff791f920 in TagLib::ID3v2::Tag::Tag (this=0x618680, file=0x617cd0, tagOffset=46, 
    factory=0x7ffff7dda9c0 <TagLib::ID3v2::FrameFactory::factory>)
    at /home/cocobo/repository/taglib/taglib/mpeg/id3v2/id3v2tag.cpp:125
No locals.
#20 0x00007ffff7ab43df in TagLib::RIFF::AIFF::File::read (this=0x617cd0, readProperties=true, 
    propertiesStyle=TagLib::AudioProperties::Average)
    at /home/cocobo/repository/taglib/taglib/riff/aiff/aifffile.cpp:136
        i = 1
#21 0x00007ffff7ab2dad in TagLib::RIFF::AIFF::File::File (this=0x617cd0, 
    file=0x7fffffffde84 "taglib_aiff_alloc.aif", readProperties=true, 
    propertiesStyle=TagLib::AudioProperties::Average)
    at /home/cocobo/repository/taglib/taglib/riff/aiff/aifffile.cpp:67
No locals.
#22 0x00007ffff7b7b89c in TagLib::FileRef::create (fileName=0x7fffffffde84 "taglib_aiff_alloc.aif", 
    readAudioProperties=true, audioPropertiesStyle=TagLib::AudioProperties::Average)
    at /home/cocobo/repository/taglib/taglib/fileref.cpp:269
        it = 
        ext = {_vptr.String = 0x7ffff7dce4f0 <vtable for TagLib::String+16>, static null = {
            _vptr.String = 0x7ffff7dce4f0 <vtable for TagLib::String+16>, 
            static null = <same as static member of an already seen type>, 
            static WCharByteOrder = TagLib::String::UTF16LE, d = 0x616af0}, 
          static WCharByteOrder = TagLib::String::UTF16LE, d = 0x617da0}
#23 0x00007ffff7b77ad1 in TagLib::FileRef::FileRef (this=0x7fffffffd8d0, 
    fileName=0x7fffffffde84 "taglib_aiff_alloc.aif", readAudioProperties=true, 
    audioPropertiesStyle=TagLib::AudioProperties::Average)
    at /home/cocobo/repository/taglib/taglib/fileref.cpp:83
No locals.
#24 0x00000000004048a7 in main (argc=2, argv=0x7fffffffda58)
    at /home/cocobo/repository/taglib/examples/tagreader.cpp:41
        f = {_vptr.FileRef = 0x7ffff7dd0610 <vtable for TagLib::FileRef+16>, d = 0x404792 <main+34>}
        i = 1
#25 0x00007ffff6c76dc5 in __libc_start_main () from /lib64/libc.so.6
No symbol table info available.
#26 0x00000000004046a9 in _start ()
No symbol table info available.
(gdb) up 11
#11 0x00007ffff7a42cde in TagLib::String::copyFromLatin1 (this=0x7fffffffd300, s=0x7fff2bd8f015 "", 
    length=3402287813) at /home/cocobo/repository/taglib/taglib/toolkit/tstring.cpp:743
743       d->data.resize(length);
(gdb) list
738     // private members
739     ////////////////////////////////////////////////////////////////////////////////
740
741     void String::copyFromLatin1(const char *s, size_t length)
742     {
743       d->data.resize(length);
744
745       for(size_t i = 0; i < length; ++i)
746         d->data[i] = static_cast<uchar>(s[i]);
747     }
(gdb) print length
$1 = 3402287813
```

**Comment by TsudaKageyu** [comment]
This and #466 are similar issues come from corrupt ID3v2 tag and/or frame headers. We should add some integrity checks of the headers to fix them.

I think it's a possible option to check if the unused flags are cleared like I did in #467.
Does anyone have an idea of good checking method of the headers?

**Comment by TsudaKageyu** [comment]
Fixed by @scotchi in caa53e8de5591a75138f90cced544908d016222b. Should be closed.
